tests/pipelines/test_pipelines_zero_shot_object_detection.py

```diff
@@ -0,0 +1,206 @@
+# limitations under the License.
+
+import unittest
+
+from transformers import is_vision_available, pipeline
+from transformers.testing_utils import (
+    is_pipeline_test,
+    nested_simplify,
+    require_tf,
+    require_torch,
+    require_vision,
+    slow,
+)
+
+from .test_pipelines_common import PipelineTestCaseMeta
+
+
+if is_vision_available():
+    from PIL import Image
+else:
+
+    class Image:
+        @staticmethod
+        def open(*args, **kwargs):
+            pass
+
+
+@require_vision
+@require_torch
+@is_pipeline_test
+class ZeroShotObjectDetectionPipelineTests(unittest.TestCase, metaclass=PipelineTestCaseMeta):
+    # Deactivating auto tests since we don't have a good MODEL_FOR_XX mapping,
+    # and only OwlViTForObjectDetection would be there for now.
+    # model_mapping = {OwlViTConfig: OwlViTForObjectDetection}
+
+    # def get_test_pipeline(self, model, tokenizer, feature_extractor):
+    #     object_detector = pipeline("zero-shot-object-detection", model="hf-internal-testing/tiny-random-owlvit")
+
+    #     return object_detector, ["./tests/fixtures/tests_samples/COCO/000000039769.png"]
+
+    # def run_pipeline_test(self, object_detector, examples):
+    #     outputs = object_detector(
+    #         Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png"),
+    #         candidate_labels=["cat", "remote", "couch"],
+    #         threshold=0.0,
+    #     )
+
+    #     self.assertGreater(len(outputs), 0)
+    #     for detected_object in outputs:
+    #         self.assertEqual(
+    #             detected_object,
+    #             {
+    #                 "score": ANY(float),
+    #                 "label": ANY(str),
+    #                 "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+    #             },
+    #         )
+
+    #     batch = [
+    #         Image.open("./tests/fixtures/tests_samples/COCO/000000039769.png"),
+    #         "http://images.cocodataset.org/val2017/000000039769.jpg",
+    #     ] * 4
+    #     batch_outputs = object_detector(batch, candidate_labels=["cat", "remote", "couch"], threshold=0.0)
+
+    #     self.assertEqual(len(batch), len(batch_outputs))
+
+    #     for outputs in batch_outputs:
+    #         self.assertGreater(len(outputs), 0)
+    #         for detected_object in outputs:
+    #             self.assertEqual(
+    #                 detected_object,
+    #                 {
+    #                     "score": ANY(float),
+    #                     "label": ANY(str),
+    #                     "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+    #                 },
+    #             )
+
+    @require_tf
+    @unittest.skip("Zero Shot Object Detection not implemented in TF")
+    def test_small_model_tf(self):
+        pass
+
+    @require_torch
+    @unittest.skip("Using hf-internal-testing/tiny-random-owlvit throws error")
```

Comment:
The reason of skipping this test is `hf-internal-testing/tiny-random-owlvit` throws the following error.

Code to reproduce bug:
```
from transformers import (
    MODEL_FOR_OBJECT_DETECTION_MAPPING,
    AutoFeatureExtractor,
    AutoTokenizer,
    AutoModelForObjectDetection,
    ZeroShotObjectDetectionPipeline,
    is_vision_available,
    pipeline,
)
from transformers.testing_utils import (
    is_pipeline_test,
    nested_simplify,
    require_tf,
    require_torch,
    require_vision,
    slow,
)

from PIL import Image

model_id = "hf-internal-testing/tiny-random-owlvit"

model = AutoModelForObjectDetection.from_pretrained(model_id)
feature_extractor = AutoFeatureExtractor.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id)

object_detector = ZeroShotObjectDetectionPipeline(
    model=model, feature_extractor=feature_extractor, tokenizer=tokenizer
)

outputs = object_detector(
    "./tests/fixtures/tests_samples/COCO/000000039769.png",
    candidate_labels=["cat", "remote", "couch"],
    threshold=0.0,
)
```

Traceback along with some warning

```
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Some weights of OwlViTForObjectDetection were not initialized from the model checkpoint at hf-internal-testing/tiny-random-owlvit and are newly initialized: ['class_head.logit_scale.bias', 'class_head.dense0.weight', 'box_head.dense0.weight', 'box_head.dense1.weight', 'layer_norm.bias', 'class_head.dense0.bias', 'class_head.logit_shift.bias', 'box_head.dense1.bias', 'class_head.logit_scale.weight', 'box_head.dense2.weight', 'layer_norm.weight', 'box_head.dense0.bias', 'box_head.dense2.bias', 'class_head.logit_shift.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Traceback (most recent call last):
  File "/home/azureuser/cloudfiles/code/Users/asahu.cool/Work/Repos/transformers/test.py", line 34, in <module>
    threshold=0.0,
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/zero_shot_object_detection.py", line 64, in __call__
    return super().__call__(images, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/base.py", line 1074, in __call__
    return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/base.py", line 1081, in run_single
    model_outputs = self.forward(model_inputs, **forward_params)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/base.py", line 990, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/pipelines/zero_shot_object_detection.py", line 94, in _forward
    outputs = self.model(**model_inputs)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1384, in forward
    (pred_logits, class_embeds) = self.class_predictor(image_feats, query_embeds, query_mask)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1256, in class_predictor
    (pred_logits, image_class_embeds) = self.class_head(image_feats, query_embeds, query_mask)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1149, in forward
    pred_logits = torch.einsum("...pd,...qd->...pq", image_class_embeds, query_embeds)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/functional.py", line 360, in einsum
    return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
RuntimeError: einsum(): operands do not broadcast with remapped shapes [original->remapped]: [1, 576, 8]->[1, 576, 1, 8] [1, 3, 16]->[1, 1, 3, 16]
```

Reply:
Hi @alaradirik . [hf-internal-testing/tiny-random-owlvit](https://huggingface.co/hf-internal-testing/tiny-random-owlvit/tree/main) is present on hub

Reply:
My mistake, sorry about that! 

The error is most likely due to calling feature extractor and tokenizer separately instead of using OwlViTProcessor, which makes sure that the `input_ids` has the correct shape.

Reply:
Hi @alaradirik , it seems `hf-internal-testing/tiny-random-owlvit` isn't even working with the demo notebook code.

**Code to reproduce bug**
`from transformers import OwlViTProcessor, OwlViTForObjectDetection
from PIL import Image
import os
import skimage
import numpy as np
from PIL import Image
import torch


image = skimage.data.astronaut()
image = Image.fromarray(np.uint8(image)).convert("RGB")
text_queries = ["human face", "rocket", "nasa badge", "star-spangled banner"]
model_id = "hf-internal-testing/tiny-random-owlvit"

if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

model = OwlViTForObjectDetection.from_pretrained(model_id)
processor = OwlViTProcessor.from_pretrained(model_id)

inputs = processor(text=text_queries, images=image, return_tensors="pt").to(device)

model = model.to(device)
model.eval()

with torch.no_grad():
    outputs = model(**inputs)`
    
**Error**
`WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Some weights of OwlViTForObjectDetection were not initialized from the model checkpoint at hf-internal-testing/tiny-random-owlvit and are newly initialized: ['box_head.dense0.bias', 'box_head.dense2.bias', 'box_head.dense1.weight', 'box_head.dense1.bias', 'class_head.logit_shift.bias', 'layer_norm.bias', 'layer_norm.weight', 'class_head.logit_scale.bias', 'class_head.logit_scale.weight', 'class_head.dense0.weight', 'box_head.dense2.weight', 'class_head.dense0.bias', 'box_head.dense0.weight', 'class_head.logit_shift.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Traceback (most recent call last):
  File "/home/azureuser/cloudfiles/code/Users/asahu.cool/Work/Repos/transformers/test.py", line 46, in <module>
    outputs = model(**inputs)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1384, in forward
    (pred_logits, class_embeds) = self.class_predictor(image_feats, query_embeds, query_mask)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1256, in class_predictor
    (pred_logits, image_class_embeds) = self.class_head(image_feats, query_embeds, query_mask)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/mnt/batch/tasks/shared/LS_root/mounts/clusters/devmachine/code/Users/asahu.cool/Work/Repos/transformers/src/transformers/models/owlvit/modeling_owlvit.py", line 1149, in forward
    pred_logits = torch.einsum("...pd,...qd->...pq", image_class_embeds, query_embeds)
  File "/anaconda/envs/transformers_hf/lib/python3.7/site-packages/torch/functional.py", line 360, in einsum
    return _VF.einsum(equation, operands)  # type: ignore[attr-defined]
RuntimeError: einsum(): operands do not broadcast with remapped shapes [original->remapped]: [1, 576, 8]->[1, 576, 1, 8] [1, 4, 16]->[1, 1, 4, 16]`


Seeing the error it seems the query and image embeddings are having different dimensions - 16 vs 8 respectively

Reply:
Hi @sahamrit, thanks for bringing this up. @lewtun could you take a look at `hf-internal-testing/tiny-random-owlvit`?

Reply:
Hi @alaradirik , @lewtun . Any updates on this? I wanted to close this PR by this week since am on vacation next week.

Reply:
Hey @sahamrit I see the problem is that the test model `hf-internal-testing/tiny-random-owlvit` is missing the object detection head. 

We're currently overhauling these tiny models - @ydshieh would the best course of action be for me to create a `hf-internal-testing/tiny-random-owlvit-object-detection` model for just this task?

Reply:
@lewtun Yes!

Reply:
OK great, I'll add this right away and report back when it's up :)

Reply:
Sorry for the delay - had some issues shrinking the model correctly. 

You can use this ID for the tests: https://huggingface.co/hf-internal-testing/tiny-random-owlvit-object-detection